File: lab5/code/lowpass.m
fs = 1600;
fc = 400;
N = 39;

[h_n] = LPF(fc, fs, N);

n = (-N + 1)/2 : (N - 1)/2;
stem(n, h_n, 'b', 'Marker', 'o', 'LineWidth', 1.5);
title('LPF Impulse Response with Hamming Window');
xlabel('n');
ylabel('Amplitude');
grid on;
legend('h[n]');

fvtool(h_n, 1, 'Fs', fs);


function [h_n] = LPF(fc,fs,N)
n = (-N + 1)/2 : (N - 1)/2;
omega_c = pi / 2;
hd_n = sin(omega_c * n) ./ (pi * n);
hd_n(N - (N - 1)/2) = omega_c / pi;
w_n = 0.54 - 0.46 * cos(2 * pi * n / (N - 1));
h_n = hd_n .* w_n;
end

File: lab5/code/lowpass.c
#include <stdio.h>
#include <math.h>

#define N 39  
#define PI 3.14159265358979323846

void hammingWindow(float h[]) {
    for (int n = 0; n < N; n++) {
        float window = 0.54 - 0.46 * cos(2 * PI * n / (N - 1));
        h[n] *= window;
    }
}

void LPF(float fc, float fs, float h[]) {
    float omega_c = 2 * PI * fc / fs;

    for (int n = -N / 2; n <= N / 2; n++) {
        if (n == 0) {
            h[n + N / 2] = omega_c / PI;
        } else {
            h[n + N / 2] = sin(omega_c * n) / (PI * n);
        }
    }
    hammingWindow(h);
}

int main() {
    float fs = 6000;   
    float fc = 400;
    float h[N];

    LPF(fc, fs, h);

    printf("Impulse Response after Hamming Window:\n");
    for (int i = 0; i < N; i++) {
        printf("%f\n", h[i]);
    }

    return 0;
}
